### PR TITLE
Small cleanup of declare_variable in declare.ml and comAssumption.ml

### PR DESCRIPTION
**Pull request by herbelin**
The PR does the following:
- it aligns `Declare.declare_variable` to the style of `Declare.declare_constant`:
   - use in `declare_variable` the "variable" equivalent to `constant_entry`, that is `variable_declaration`
   - include thus support for local definitions
- it aligns `ComAssumption.declare_variable` on the style of `ComAssumption.declare_axiom`; it also gets rid of a useless `Loc.t`

This covers the most important source of API incompatibility of #13445.

Synchronous overlays:
- smtcoq/smtcoq#132
- MetaCoq/metacoq#1029
- LPCIC/coq-elpi#561

### CODE REVIEW

File: vernac/declare.mli
Comment (SkySkimmer) [inline vernac/declare.mli:391]: something like 
```suggestion
  | SectionLocalAssum of {
      typ:Constr.types;
      impl:Glob_term.binding_kind ;
      univs:UState.named_universes_entry
      }
```
is nicer IMO (I didn't count the spaces as I wrote it in the github UI)

Comment (herbelin) [reply vernac/declare.mli]: Adopted (as well as for `SectionLocalDef` then).

PS: not relaunching full ci as it was already found green.

**Comment by SkySkimmer** [comment]
TBH I'm not sure this PR is all that useful, the current type of declare_variable looks fine to me.

**Comment by herbelin** [comment]
The motivation is to have a more regular API (compare to declare_constant). Also, proof_entry and methods to build a proof_entry are already exported so that `declare_variable (SectionLocalDef ...)` is already doable via `declare_entry`.

I'm not opposed to alternative APIs (hiding proof_entry?) but assuming proof_entry exported, the API obtained from the current PR looks more regular to me.

**Comment by herbelin** [comment]
@SkySkimmer: Is your concern that you find there are already too many ways to define a `Let`?

**Comment by herbelin** [comment]
Ping @coq/vernac-maintainers

File: vernac/declare.mli
Comment (proux01) [reply vernac/declare.mli:391]: @herbelin this question is still waiting for your answer.

Are we sure we want to declare definitions via `declare_variable`? or do we want to do it via `declare_definition` as done for `Let`?

@ejgallego according to the comments, it seems you had some opinions on this API but I'm not sure I completely got it, in particular the precise meaning of "higher-level declare API" in the comment of `declare_entry`.

**Comment by proux01** [comment]
@herbelin maybe this needs a discussion during a forthcoming Coq call? the discussion doesn't seem to make much progress here.

> @SkySkimmer: Is your concern that you find there are already too many ways to define a Let?

Personally, I would find it better to avoid introducing a new way to do it, at least without removing the current way.

**Comment by herbelin** [comment]
Side question: do we really need to forbid `Let` defined by proofs? Wouldn't we have a more uniform system by allowing it. Isn't the problem more to keep a transparency information in the `named_context` that forbidding defining `Let` with proofs?

**Comment by herbelin** [comment]
> maybe this needs a discussion during a forthcoming Coq call? 

OK, I added an item to the next call.

**Comment by proux01** [comment]
This was agreed during today call: https://github.com/coq/coq/wiki/Coq-Call-2024-02-19 so let's rerun CI to be safe and merge.

**Review by proux01 — APPROVED** [review]

**Comment by proux01** [comment]
The tactician failure is unrelated (I can reproduce it on master), Cc @LasseBlaauwbroek maybe due to that unmerged overlay: https://github.com/coq-tactician/coq-tactician/pull/79

**Comment by proux01** [comment]
And bedrock2 seem to be the usual timeout, let's merge

@coqbot merge now

**Pull request merged**
